homeassistant/components/onvif/__init__.py

```diff
@@ -118,6 +120,7 @@ async def async_populate_options(hass, entry):
     options = {
         CONF_EXTRA_ARGUMENTS: DEFAULT_ARGUMENTS,
         CONF_RTSP_TRANSPORT: RTSP_TRANS_PROTOCOLS[0],
+        CONF_SNAPSHOT_AUTH: HTTP_DIGEST_AUTHENTICATION,
```

Comment:
I'm confused. If this is part of the options, how can it pass the config flow when the camera isn't added yet?

The options are set after the fact, right? So how would it pass the initial integration setup?

Reply:
that's a good question. Technically, this is optional - the integration is working even without snapshots. I don't think it's different than `CONF_RTSP_TRANSPORT` which is also an optional connection detail, but I don't know the integration well enough to answer this.

Reply:
I'd say it should stay as an option flow vs config flow during initial setup because most people won't know the answer here, and depending on how the vendor implemented it it could differ. Having Digest as the default makes it something most people won't need to worry about, but for those that need basic auth, they don't have to remove and re-add the entire integration to change it.

Reply:
Basically, we don't actually try to fetch the snapshot until after the config flow passes, the integration is loaded and the camera is viewed from the frontend, so it's not required for initial setup.

Reply:
So, actually, I still think this still doesn't belong in the options.

> because most people won't know the answer here

Yes, I agree, so why don't we try and auto-detect it? When setting up the integration, try fetching it using the Digest method, if that fails, try the basic auth. If all fails...?

Reply:
I like that.

Reply:
I'll try to draft something up.
In my case, it doesn't explicitly fail, it just times out. I guess I'll try with a small timeout, and fall back to basic auth on timeout error, and on 401.
Not sure what do to with other types of failures though, because snapshot failure is an expected condition that has a fallback in the code.

Reply:
@OnFreund - I think @frenck's suggestion is to add the Basic Auth fallback inbetween the Digest Auth and FFMPEG fallback so it's 3 tiered.
1) Try Digest
2) Try Basic
3) Use FFMPEG

Reply:
> When setting up the integration, try fetching it using the Digest method, if that fails, try the basic auth. If all fails...?

I read this as config flow, but maybe I got it wrong.
I really prefer the config flow approach so, since (at least in my case), there's a timeout involved, so I'd rather have a few of those as possible.

Reply:
Also, NVR can have dozens of cameras, and if it doesn't work for one, it doesn't work for all. Keeping the logic at the entity level will make it hard to avoid having the timeout in each one.

Reply:
OK, this is way more difficult than I anticipated. To get the snapshot uri, I need an onvif device, and to get an onvif device I need a saved config entry (the dict returned from `async_create_entry` is not enough).
On the other hand, doing this within the camera entity can leads to all of timeouts.
Right now I'm leanings towards testing this in `async_setup_entry` in `__init__.py` in case that value doesn't exist.